src/c++/perf_analyzer/test_request_rate_manager.cc

```diff
@@ -312,6 +312,104 @@ class TestRequestRateManager : public TestLoadManagerBase,
+      std::vector<std::vector<int32_t>>& expected_values,
+      bool expect_init_failure, bool expect_thread_failure)
+  {
+    params_.user_data = {json_str};
+
+    TestCustomDataSetup(tensors);
+
+    if (expect_init_failure) {
+      REQUIRE_THROWS_AS(
+          InitManager(
+              params_.string_length, params_.string_data, params_.zero_input,
+              params_.user_data, params_.start_sequence_id,
+              params_.sequence_id_range, params_.sequence_length,
+              params_.sequence_length_specified,
+              params_.sequence_length_variation),
+          PerfAnalyzerException);
+      return;
+    } else {
+      REQUIRE_NOTHROW(InitManager(
+          params_.string_length, params_.string_data, params_.zero_input,
+          params_.user_data, params_.start_sequence_id,
+          params_.sequence_id_range, params_.sequence_length,
+          params_.sequence_length_specified,
+          params_.sequence_length_variation));
+    }
+
+    auto thread_status = TestCustomDataRunThread();
+
+    if (!thread_status.IsOk() && !expect_thread_failure) {
+      std::cout << "Unexpected thread failure: " << thread_status.Message()
+                << std::endl;
+    }
+
+    REQUIRE(thread_status.IsOk() != expect_thread_failure);
+    if (!thread_status.IsOk()) {
+      return;
+    }
+
+    auto recorded_inputs{stats_->recorded_inputs};
+    std::vector<std::vector<int32_t>> recorded_values;
+
+    // Convert the recorded inputs into values, for both shared memory and non
+    // shared memory cases
+    //
+    if (params_.shared_memory_type != SharedMemoryType::NO_SHARED_MEMORY) {
+      auto recorded_memory_regions =
+          std::dynamic_pointer_cast<MockInferDataManagerShm>(
+              infer_data_manager_)
+              ->mocked_shared_memory_regions;
+      for (auto recorded_input : recorded_inputs) {
+        std::vector<int32_t> recorded_value;
+        for (auto memory_label : recorded_input) {
+          auto itr =
+              recorded_memory_regions.find(memory_label.shared_memory_label);
+          if (itr == recorded_memory_regions.end()) {
+            std::string err_str = "Test error: Could not find label " +
+                                  memory_label.shared_memory_label +
+                                  " in recorded shared memory";
+            REQUIRE_MESSAGE(false, err_str);
+          } else {
+            for (auto val : itr->second) {
+              recorded_value.push_back(val);
+            }
+          }
+        }
+        recorded_values.push_back(recorded_value);
+      }
+    } else {
+      for (auto recorded_input : recorded_inputs) {
+        std::vector<int32_t> recorded_value;
+        for (auto val : recorded_input) {
+          recorded_value.push_back(val.data);
+        }
+        recorded_values.push_back(recorded_value);
+      }
+    }
+
+    // Check that results are exactly as expected
+    REQUIRE(recorded_values.size() >= expected_values.size());
```

Comment:
why isnt this strictly equal?

Reply:
Due to this:
https://github.com/triton-inference-server/client/pull/259#discussion_r1132971350
We turn the thread on for some set amount of time, but based on thread scheduling it may issue more requests before it is turned off.

Reply:
got it. thank you